core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+	}
+
+	if objectType != "" {
+		query["metadata.objectType"] = objectType
+		if objectID != "" {
+			query["metadata.objectID"] = objectID
+		}
+	}
+
+	if destinationType != "" {
+		var subquery []map[string]interface{}
+		if destinationID == "" {
+			subquery = []map[string]interface{}{
+				map[string]interface{}{"metadata.destinationType": destinationType},
+				map[string]interface{}{"metadata.destinationsList": map[string]interface{}{"$allMatch": map[string]interface{}{"$regex": destinationType + ":*"}}},
+			}
+		} else {
+			subquery = []map[string]interface{}{
+				map[string]interface{}{"metadata.destinationType": destinationType, "metadata.destinationID": destinationID},
+				map[string]interface{}{"metadata.destinationsList": map[string]interface{}{"$allMatch": map[string]interface{}{
+					"$eq": destinationType + ":" + destinationID}},
+				}}
+		}
+		query["$or"] = subquery
+	}
+
+	if noData != nil {
+		query["metadata.noData"] = *noData
+	}
+
+	if expirationTimeBefore != "" {
+		subquery := map[string]interface{}{
+			"$ne":  "",
+			"$lte": expirationTimeBefore,
+		}
+		query["metadata.expiration"] = subquery
+	}
+
+	finalQuery := map[string]interface{}{"selector": query}
+
+	if err := store.findAll(finalQuery, &result); err != nil {
+		return nil, &Error{fmt.Sprintf("Failed to fetch the objects. Error: %s.", err)}
+	}
+
+	metaDatas := make([]common.MetaData, len(result))
+	for i, r := range result {
+		metaDatas[i] = r.MetaData
+	}
+	return metaDatas, nil
+}
+
+// RetrieveConsumedObjects returns all the consumed objects originated from this node
+// ESS only API
+func (store *CouchStorage) RetrieveConsumedObjects() ([]common.ConsumedObject, common.SyncServiceError) {
+	return nil, nil
+}
+
+// RetrieveObjectData returns the object data with the specified parameters
+func (store *CouchStorage) RetrieveObjectData(orgID string, objectType string, objectID string) (io.Reader, common.SyncServiceError) {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	attachment, err := store.getAttachment(id)
+	if err != nil {
+		switch err {
+		case notFound:
+			return nil, nil
+		default:
+			return nil, &Error{fmt.Sprintf("Failed to open file to read the data. Error: %s.", err)}
+		}
+	}
+
+	data, err := ioutil.ReadAll(attachment.Content)
+	defer attachment.Content.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes.NewReader(data), nil
+}
+
+// CloseDataReader closes the data reader if necessary
+func (store *CouchStorage) CloseDataReader(dataReader io.Reader) common.SyncServiceError {
+	return nil
+}
+
+// ReadObjectData returns the object data with the specified parameters
+func (store *CouchStorage) ReadObjectData(orgID string, objectType string, objectID string, size int, offset int64) ([]byte, bool, int, common.SyncServiceError) {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	attachment, err := store.getAttachment(id)
+	if err != nil {
+		if err == notFound {
+			return nil, true, 0, notFound
```

Comment:
Should this be `return nil, true, 0, &common.NotFound{}`?

Reply:
is `&common.NotFound{}` not the same as [this ](https://github.com/open-horizon/edge-sync-service/blob/dde1bef314097c720c0026c127f4854b1cc82ee7/core/storage/storage.go#L319) ?

Reply:
ah yes you're right